rfcs/0018-glides-trainsheet-events.md

```diff
@@ -0,0 +1,361 @@
+}
+
+or
+
+{
+  "username": "pswartz"
+}
+```
+It is represented as an object to provide future extensibility if needed.
+
+### Car
+Cars (individual parts of a train consist) will be described by their 4 digit number.
+```json
+{
+  "carNumber": "3800"
+}
+```
+
+### Operator
+Operators will be described by their badge number:
+```json
+{
+  "badgeNumber": "123"
+}
+```
+
+### Optional T
+A value that can either be an instance of T, or `null`. For example, Optional Car can be
+```json
+{ "carNumber": "3800"}
+# or 
+null
+```
+
+### Time
+A [GTFS Time](https://github.com/google/transit/blob/master/gtfs/spec/en/reference.md#field-types).
+> Time in the HH:MM:SS format (H:MM:SS is also accepted). The time is measured from "noon minus 12h" of the service day (effectively midnight except for days on which daylight savings time changes occur). For times occurring after midnight, enter the time as a value greater than 24:00:00 in HH:MM:SS local time for the day on which the trip schedule begins.  
+> _Example: `14:30:00` for 2:30PM or `25:35:00` for 1:35AM on the next day._
+### Trip Key
+For scheduled trips, Glides does not have the same trip ID used by RTR. For added trips, while an ID is provided, it can be the case that an added trip from the Glides perspective is matches a GTFS trip, in which case RTR will us the GTFS trip ID in publishing the information.
+
+In order to reduce message duplication, a Trip Key is used to identify both added and scheduled trips.
+
+**Added Trip**
+- `serviceDate` (RFC3999 date): the service date for the trip
+- `id` (string): unique value representing this trip for the given `serviceDate`
+```json
+`{"serviceDate": "2023-01-20", "id": "ADDED-123"}
+```
+**Scheduled Trip (departure time from origin)
+- `serviceDate` (RFC3999 date): the service date for the trip
+- `departureTime` (Time): when the vehicle is scheduled to depart the given stop
+- `startStop` (string): GTFS stop ID the trip is departing from (either `location_type` 0 or 1)
+```json
+`{"serviceDate": "2023-01-20", "departureTime": "25:01:02", "startStop": "place-lake"}
+```
+**Scheduled Trip (arrival time at destination)
+- `serviceDate` (RFC3999 date): the service date for the trip
+- `arrivalTime` (Time): when the vehicle is scheduled to arrive at the given stop
+- `endStop` (string): GTFS stop ID the trip is ending at (either `location_type` 0 or 1)
+```json
+`{"serviceDate": "2023-01-20", "arrivalTime": "09:10:11", "endStop": "place-lake"}
+```
+**Scheduled Trip (known GTFS trip ID)
+- `serviceDate` (RFC3999 date): the service date for the trip
+- `gtfsId` (string): GTFS trip ID
+```json
+`{"serviceDate": "2023-01-20", "gtfsId": "123456790"}
+```
+
+### Trip Type
+When adding a trip, there are three kinds of trips which can be added:
+- `"arrival"`: a trip arriving to `endStop` at `time`
+- `"departure"`: a trip departing from `startStop` at `time`
+- `"roundTrip"`: a trip departing at `time` and expected to return to `startStop`
```

Comment:
It is a single action in the UI to add a round trip. And the second trip won't really have any information aside from "it's the return of the previous trip".

But I do think the way the event exactly mirrors the UI of Glides makes it worse at representing the new data. Someone has to do the "the inspector clicked round trip in Glides, and that means making 2 trips" logic, and I think it makes more sense for that logic to be in Glides producing multiple trips, than for the consumer to have to say "we treat round trips like this because that's how the Glides UI works".

The endpoint-perspective "arrival/departure" also seems like a Glides-specific concern leaking into the event. What if we later allow a dispatcher to add a trip. Do we pretend they're at a terminal so they can choose arrival or a departure? Would we need to add a second "added a trip without a perspective" event?

New idea: adding a trip adds a list of trips, so that adding a round trip returns a list of two associated trips within a single event. Both of the trips have 4 perspectiveless fields "start/end time" and "start/end station". Fields that aren't manually specified aren't filled. The current arrival/departure/roundTrip would create events with fields:
```jsonc
// departure
{
  "data": {
    "author": {
      "username": "ainspector",
      "badgeNumber": "123",
      "stop": "place-matt"
    },
    trips: [
      {
        "tripKey": {/*generated*/},
        "startTime": "10:00:00",
        "startStation": "place-matt",
      },
    ]
  }
}
// arrival
{
  "data": {
    "author": {/*as above*/},
    trips: [
      {
        "tripKey": {/*generated*/},
        "endTime": "10:00:00",
        "endStation": "place-matt",
      },
    ]
  }
}
// roundTrip
{
  "data": {
    "author": {/*as above*/},
    trips: [
      {
        "tripKey": {/*generated*/},
        "startTime": "10:00:00",
        "startStation": "place-matt",
      },
      {
        "tripKey": {/*generated*/},
        "endStation": "place-matt",
        // no times, but it's assumed to be immediately after the previous trip because of the ordering in the list.
      },
    ]
  }
}
```

Reply:
My reasoning for including this as a single item was to capture the inspector's input in a way where RTR could generate a reverse prediction without any additional information. In order to have two separate trips, we'd need a time for when the train is expected to start the reverse trip, which to me is an RTR responsibility rather than a Glides responsibility (since it's a prediction about the future, over which the inspector has basically no influence).

I do see what you're saying about not being able to edit or refer to the reverse trip; I'll include something for that.

Reply:
Added `previousTripKey` and `nextTripKey`.